### ISSUE TITLE
Inconsistency in types returned by Object.assign

### ISSUE THREAD

**Issue by bjouhier** [microsoft/TypeScript#13912]
**TypeScript Version:**  2.1.5

**Code**

```ts
const foo = Object.assign({}, {
    f() { return 1; }
});
// next line does not give error because foo is typed as any :-(
foo.g();

const bar = {
    f() { return 1; }
}
const zoo = Object.assign({}, bar);
// but next line does give error :-)
zoo.h();
```

**Expected behavior:**
foo.ts(5,5): error TS2339: Property 'g' does not exist on type '{} & { f(): number; }'.
foo.ts(12,5): error TS2339: Property 'h' does not exist on type '{} & { f(): number; }'.

**Actual behavior:**
foo.ts(12,5): error TS2339: Property 'h' does not exist on type '{} & { f(): number; }'.

**Comment by mhegazy** [comment]
Though not the same issue, but should be covered by https://github.com/Microsoft/TypeScript/issues/11100

**Comment by RyanCavanaugh** [comment]
This has been fixed
